pkg/logql/probabilistic_evaluator.go

```diff
@@ -0,0 +1,325 @@
+package logql
+
+import (
+	"context"
+	"fmt"
+	"math"
+	"sort"
+	"time"
+
+	"github.com/go-kit/log"
+	"github.com/opentracing/opentracing-go"
+	"github.com/pkg/errors"
+	"github.com/prometheus/prometheus/model/labels"
+	"github.com/prometheus/prometheus/promql"
+	promql_parser "github.com/prometheus/prometheus/promql/parser"
+
+	"github.com/grafana/dskit/tenant"
+
+	"github.com/grafana/loki/pkg/logql/sketch"
+	"github.com/grafana/loki/pkg/logql/syntax"
+	"github.com/grafana/loki/pkg/logqlmodel"
+	"github.com/grafana/loki/pkg/util"
+	"github.com/grafana/loki/pkg/util/validation"
+)
+
+type ProbabilisticEvaluator struct {
+	DefaultEvaluator
+	logger log.Logger
+}
+
+// NewDefaultEvaluator constructs a DefaultEvaluator
+func NewProbabilisticEvaluator(querier Querier, maxLookBackPeriod time.Duration) Evaluator {
+	d := NewDefaultEvaluator(querier, maxLookBackPeriod)
+	p := &ProbabilisticEvaluator{DefaultEvaluator: *d}
+	return p
+}
+
+func (p *ProbabilisticEvaluator) StepEvaluator(
+	ctx context.Context,
+	nextEv SampleEvaluator,
+	expr syntax.SampleExpr,
+	q Params,
+) (StepEvaluator, error) {
+
+	switch e := expr.(type) {
+	case *syntax.VectorAggregationExpr:
+		if e.Operation != syntax.OpTypeTopK {
+			return p.DefaultEvaluator.StepEvaluator(ctx, nextEv, expr, q)
+		}
+		return p.newProbabilisticVectorAggEvaluator(ctx, nextEv, e, q)
+	default:
+		return p.DefaultEvaluator.StepEvaluator(ctx, nextEv, expr, q)
+	}
+}
```

Comment:
Would we eventually want to have a single engine that just executes a query in the way we implement it to, instead of two separate engines? 

It feels a bit odd to me to be doing the check here in the creation of the step evaluator rather than higher up somewhere we would decide to pass a query to one of the two engines we have in this PR. I think we'd also want a type that contains a list of `probabilisticQueries` similar to how we have `shardableQueries` at the moment. Then the list of things that are (or can be) executed in a probabilistic manner is centralized in one place, and we just check `syntax.<expression type>` against that list.

Reply:
`StepEvaluator` is called recursively. It's a little hidden but in the original engine 

`DefaultEvaluator.StepEvaluator` calls `vectorAggEvaluator(ctx, nextEv, e, q)` which then calls `ev.StepEvaluator(ctx, ev, expr.Left, q)`. Where `ev` is `nextEv` which is `DefaultEvaluator`.

So we have to slip in the probabilistic evaluator and then switch back to the original. However, I agree. This should happen outside of the evaluators. Ideally the evaluators would not construct the step evaluators but something out side that builds up the chain of evaluators.